proposals/0000-bytearray-literals.rst

```diff
@@ -0,0 +1,345 @@
+
+Ordinary strings, like ``"hello``, ``"Юникод"``, ``"\NUL"`` are then desugared as
+
+.. code-block:: haskell
+
+  unpackCString# "hello"#
+  unpackCStringUtf8# "\208\174\208\189\208\184\208\186\208\190\208\180"#
+  unpackCString# "\192\128"#
+
+Current representations are null-terminated, using *Modified UTF-8* for literals
+containing the NUL character.
+
+Proposed Change Specification
+-----------------------------
+
+Primitive string literals
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The new syntax for the primitive string literals uses a suffix after the magic hash:
+
+.. code-block:: haskell
+
+  "foo"#[abu]encoding
+
+where
+
+* the first letter specifies the type of the literal:
+
+  * ``a`` - ``Addr#``
+  * ``b`` - ``ByteArray#``
+  * ``u`` - Unboxed sum: ``(# Int#, Addr# #)``
+
+  The syntax is extensible, if there will be need for new types of primitive strings.
+
+* the rest of the suffix is the encoding, which is one of following
+
+  * ``Utf8``
+  * ``Ascii``
+  * ``Latin1``
+  * ``Bytes``, which is the synonym for ``Latin1``, but captures the intent better.
+  * ``Utf16``, ``Utf16le``, ``Utf16be``, ``Utf32``, ``Utf32le``, ``Utf32be``
+
+  Encoding is matched case-insensitively. Encoding may be omitted, in that case ``Utf8`` is assumed.
+
+If GHC encounters a literal which cannot be faithfully represented in the target encoding an error will be raised. For example
+
+.. code-block:: haskell
+
+  "ÄETSÄ"#bAscii
+
+will fail with an error:
+
+::
+
+  primitive ASCII string literal must contain only characters <= '\xFF
+
+Error will be raised also if invalid Unicode surrogates are present, as
+it's an occurrence of the error case above.
+
+.. code-block:: haskell
+
+  "\xd8000"#bUtf16
+
+will fail with an error
+
+::
+
+  invalid character or shift sequence at the end of the buffer.
+
+The representations will still be NUL terminated. There is little reason
+to not to add additional byte. Therefore
+
+    "\NUL\NUL"#aASCII
+
+would produce a *three* bytes: 0x00 0x00 0x00.
+
+To summarize, there are various errors which may be raised,
+
+- Characters not representable in character set (e.g. a Unicode character in an ASCII string)
+- An unencodable codepoint (e.g. a surrogate codepoint in a UTF-8 string)
+- An invalid sequence (e.g. an unpaired surrogate in a UTF-16 string)
+
+
+*Note:* If you like to play with incorrect encodings try:
+
+::
+
+  perl -e 'print "a\x00b\x00c\x00\x00\xd8"'|iconv -f utf16le -t utf8|hexdump -C
+
+You may append two bytes to the input, try to make correct surrogate pair!
+
+
+Primitive string without modifier
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The current primitive string
+
+.. code-block:: haskell
+
+  "hello#
+
+will mean the same as
+
+.. code-block:: haskell
+
+  "hello"#aLatin1
+
+Therefore, code with current primitive strings won't break.
+
+*Unresolved:* Should there be a (``-Wall``) warning in this case, asking user to be explicit?
```

Comment:
I'm not sure I see the advantage. Implicit is fine given this syntax has a long history here and avoids the nasty compatibility issues that warning generally brings.